mypy/stubtest.py

```diff
@@ -1440,7 +1444,20 @@ def is_read_only_property(runtime: object) -> bool:
 
 def safe_inspect_signature(runtime: Any) -> inspect.Signature | None:
     try:
-        return inspect.signature(runtime)
+        try:
+            return inspect.signature(runtime)
+        except ValueError:
+            if (
+                hasattr(runtime, "__text_signature__")
+                and "<unrepresentable>" in runtime.__text_signature__
+            ):
+                # Try to fix up the signature. Workaround for
+                # https://github.com/python/cpython/issues/87233
+                sig = runtime.__text_signature__.replace("<unrepresentable>", "...")
```

Comment:
This doesn't lead to great error messages, due to the fact that the repr of `...` is -- surprisingly! -- `"Ellipsis"`:

```
error: sqlite3.dbapi2.Connection.execute is inconsistent, stub argument "parameters" should be positional-only (rename with a leading double underscore, i.e. "__parameters")
Stub: in file stdlib\sqlite3\dbapi2.pyi:355
def (self: sqlite3.dbapi2.Connection, sql: builtins.str, parameters: Union[_typeshed.SupportsLenAndGetItem[Union[builtins.str, _collections_abc.Buffer, builtins.int, builtins.float, None, Any]], typing.Mapping[builtins.str, Union[builtins.str, _collections_abc.Buffer, builtins.int, builtins.float, None, Any]]] =) -> sqlite3.dbapi2.Cursor
Runtime:
def (self, sql, parameters=Ellipsis, /)
```

How about doing something like this (the diff is relative to your PR branch)? As a bonus, it feels a little less hacky, too:

```diff
diff --git a/mypy/stubtest.py b/mypy/stubtest.py
index 3ea11e0b2..407c356d8 100644
--- a/mypy/stubtest.py
+++ b/mypy/stubtest.py
@@ -55,6 +55,14 @@ MISSING: typing_extensions.Final = Missing()
 T = TypeVar("T")
 MaybeMissing: typing_extensions.TypeAlias = Union[T, Missing]

+
+class Unrepresentable:
+    def __repr__(self) -> str:
+        return "<unrepresentable>"
+
+
+UNREPRESENTABLE: typing_extensions.Final = Unrepresentable()
+
 _formatter: typing_extensions.Final = FancyFormatter(sys.stdout, sys.stderr, False)


@@ -658,10 +666,7 @@ def _verify_arg_default_value(
                 if (
                     stub_default is not UNKNOWN
                     and stub_default is not ...
-                    # We use "..." as a hacky representation for "<unrepresentable>"
-                    # defaults on builtins, so ignore those. "..." is unlikely to be used
-                    # as a real default.
-                    and runtime_arg.default is not ...
+                    and runtime_arg.default is not UNREPRESENTABLE
                     and (
                         stub_default != runtime_arg.default
                         # We want the types to match exactly, e.g. in case the stub has
@@ -1455,7 +1460,12 @@ def safe_inspect_signature(runtime: Any) -> inspect.Signature | None:
                 # https://github.com/python/cpython/issues/87233
                 sig = runtime.__text_signature__.replace("<unrepresentable>", "...")
                 sig = inspect._signature_fromstr(inspect.Signature, runtime, sig)  # type: ignore[attr-defined]
-                return sig  # type: ignore[no-any-return]
+                assert isinstance(sig, inspect.Signature)
+                new_params = [
+                    parameter.replace(default=UNREPRESENTABLE) if parameter.default is ... else parameter
+                    for parameter in sig.parameters.values()
+                ]
+                return sig.replace(parameters=new_params)
             else:
                 raise
     except Exception:
diff --git a/mypy/test/teststubtest.py b/mypy/test/teststubtest.py
index 7050b66e1..61bba2e00 100644
--- a/mypy/test/teststubtest.py
+++ b/mypy/test/teststubtest.py
@@ -429,7 +429,10 @@ class StubtestUnit(unittest.TestCase):
         # Simulate "<unrepresentable>"
         yield Case(
             stub="def f11(text: object = None) -> None: ...",
-            runtime="def f11(text = ...): pass",
+            runtime="""
+            def f11(text=None) -> None: ...
+            f11.__text_signature__ = '(text=<unrepresentable>)'
+            """,
             error=None,
         )
```

With that change, we get better error messages from stubtest:

```
error: sqlite3.dbapi2.Connection.execute is inconsistent, stub argument "parameters" should be positional-only (rename with a leading double underscore, i.e. "__parameters")
Stub: in file stdlib\sqlite3\dbapi2.pyi:355
def (self: sqlite3.dbapi2.Connection, sql: builtins.str, parameters: Union[_typeshed.SupportsLenAndGetItem[Union[builtins.str, _collections_abc.Buffer, builtins.int, builtins.float, None, Any]], typing.Mapping[builtins.str, Union[builtins.str, _collections_abc.Buffer, builtins.int, builtins.float, None, Any]]] =) -> sqlite3.dbapi2.Cursor
Runtime:
def (self, sql, parameters=<unrepresentable>, /)
```